File: node/director/validate-reading/handler.test.js
'use strict'

const assert = require('node:assert/strict')
const { test } = require('node:test')

const handler = require('./handler')

test('normalizes a valid reading', async () => {
  const context = new TestContext()
  await handler({
    body: {
      device_id: ' pump-17 ',
      temperature_c: 42.5,
      battery_percent: 80
    }
  }, context)

  assert.equal(context.statusCode, 200)
  assert.deepEqual(context.body, {
    device_id: 'pump-17',
    temperature_c: 42.5,
    battery_percent: 80
  })
})

test('rejects a missing device ID', async () => {
  const context = new TestContext()
  await handler({ body: { temperature_c: 20 } }, context)

  assert.equal(context.statusCode, 400)
  assert.match(context.body, /device_id is required/)
})

test('rejects values outside the accepted ranges', async () => {
  const context = new TestContext()
  await handler({
    body: {
      device_id: 'pump-17',
      temperature_c: 250,
      battery_percent: 80
    }
  }, context)

  assert.equal(context.statusCode, 400)
  assert.match(context.body, /temperature_c/)
})

class TestContext {
  status (value) {
    this.statusCode = value
    return this
  }

  headers (value) {
    this.headerValues = value
    return this
  }

  succeed (value) {
    this.body = value
    return value
  }
}

File: python/director/validate-reading/handler_test.py
import json
from types import SimpleNamespace

try:
    from . import handler as h
except ImportError:
    import handler as h


def event(body):
    return SimpleNamespace(body=json.dumps(body).encode())


def test_normalizes_valid_reading():
    response = h.handle(
        event({"device_id": " pump-17 ", "temperature_c": 48.2, "battery_percent": 78}),
        {},
    )

    assert response["statusCode"] == 200
    assert response["body"]["device_id"] == "pump-17"


def test_rejects_invalid_values():
    response = h.handle(
        event({"device_id": "pump-17", "temperature_c": 250, "battery_percent": 78}),
        {},
    )

    assert response["statusCode"] == 400
    assert "temperature_c" in response["body"]


def test_rejects_invalid_json():
    response = h.handle(SimpleNamespace(body=b"not-json"), {})
    assert response["statusCode"] == 400
